solidity/contracts/Gravity.sol

```diff
@@ -80,6 +88,10 @@ contract Gravity is ReentrancyGuard {
 	// This is set once at initialization
 	bytes32 public immutable state_gravityId;
 
+	// Store vouchers for reverted cases
+	mapping(uint256 => TransferReverted) public state_RevertedVouchers;
```

Comment:
storing a hash instead of struct can maybe save some memory, not sure if it worth it?